helpers/compile/plugins/noSideEffectsPlugin.ts

```diff
@@ -0,0 +1,29 @@
+import type { Plugin } from 'esbuild'
+
+/**
+ * Plugin that forces "sideEffects": false on third party packages
+ * that do not have it specified.
+ * Needed for esbuild to be able to tree-shake those packages if
+ * they are not used.
+ *
+ * @param pattern plugin applies only to the modules matching the pattern
+ * @returns
+ */
+export function noSideEffectsPlugin(pattern: RegExp): Plugin {
+  return {
+    name: 'noSideEffectsPlugin',
+    setup(build) {
+      build.onResolve({ filter: pattern }, async (args) => {
+        if (args.pluginData?.resolved === true) {
+          return undefined
+        }
+        args.pluginData = { resolved: true }
+        const { path, ...rest } = args
+
+        const result = await build.resolve(path, rest)
+        result.sideEffects = false
+        return result
+      })
+    },
+  }
+}
```

Comment:
Isn't a "one-size-fit-all" solution like this one a bit dangerous? What if some third-party packages have side effects without specifying that? What if we used an allowlist of plugins to override, instead?

Reply:
Allowlist is exactly what this plugin is doing. Note the `pattern` parameter and the way it is used in `build.ts` - it will only apply to the modules that match pattern.

Reply:
Thank you! I had looked at it too quickly and didn't realise it